pkg/image/skopeo.go

```diff
@@ -0,0 +1,182 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package image
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+
+	"kubevirt.io/containerized-data-importer/pkg/system"
+)
+
+const dataTmpDir string = "/data_tmp"
+const whFilePrefix string = ".wh."
+
+// SkopeoOperations defines the interface for executing skopeo subprocesses
+type SkopeoOperations interface {
+	CopyImage(string, string, string, string) error
+}
+
+type skopeoOperations struct{}
+
+type manifest struct {
+	SchemaVersion int     `json:"schemaVersion"`
+	Layers        []layer `json:"layers"`   // schemaVersion v2
+	FsLayers      []layer `json:"fsLayers"` // schemaVersion v1
+}
+type layer struct {
+	Digest  string `json:"digest"`  // schemaVersion v2
+	BlobSum string `json:"blobSum"` // schemaVersion v1
+}
+
+var (
+	skopeoExecFunction = system.ExecWithLimits
+	processLimits      = &system.ProcessLimitValues{AddressSpaceLimit: maxMemory, CPUTimeLimit: maxCPUSecs}
+	// SkopeoInterface the skopeo operations interface
+	SkopeoInterface = NewSkopeoOperations()
+)
+
+// NewSkopeoOperations returns the default implementation of SkopeoOperations
+func NewSkopeoOperations() SkopeoOperations {
+	return &skopeoOperations{}
+}
+
+func (o *skopeoOperations) CopyImage(url, dest, accessKey, secKey string) error {
+	var err error
+	if len(accessKey) > 0 && len(secKey) > 0 {
+		var creds = "--src-creds=" + accessKey + ":" + secKey
+		_, err = skopeoExecFunction(processLimits, nil, "skopeo", "copy", url, dest, creds)
+	} else {
+		_, err = skopeoExecFunction(processLimits, nil, "skopeo", "copy", url, dest, "--src-tls-verify=false")
+	}
+	if err != nil {
+		return errors.Wrap(err, "could not copy image")
+	}
+	return nil
+}
+
+// ExtractTar extracts a tar file to a specified destination
+func ExtractTar(file, dest string) error {
+	_, err := skopeoExecFunction(processLimits, nil, "tar", "-xf", file, "-C", dest)
+	if err != nil {
+		return errors.Wrap(err, "could not extract image")
+	}
+	return nil
+}
+
+// CopyRegistryImage download image from registry with skopeo
+func CopyRegistryImage(url, dest, accessKey, secKey string) error {
+	skopeoDest := "dir:" + dest + dataTmpDir
+	err := SkopeoInterface.CopyImage(url, skopeoDest, accessKey, secKey)
+	if err != nil {
+		os.RemoveAll(dest + dataTmpDir)
+		return errors.Wrap(err, "Failed to download from registry")
+	}
+	err = extractImageLayers(dest)
+	if err != nil {
+		return errors.Wrap(err, "Failed to extract image layers")
+	}
+
+	// Clean temp folder
+	os.RemoveAll(dest + dataTmpDir)
+
```

Comment:
Do we want to resize like the http source as well?

Reply:
The docker image could contain any sort of the (filesystem/data/etc), so it's not necessarily an image file. Do we want to add a special behaviour just for docker images containing only disk.img?